webrender/src/device/gfx.rs

```diff
@@ -577,48 +517,47 @@ impl<B: hal::Backend> Image<B> {
         &mut self,
         device: &mut B::Device,
         cmd_pool: &mut hal::CommandPool<B, hal::queue::Graphics>,
+        staging_buffer_pool: &mut BufferPool<B>,
         rect: DeviceUintRect,
         layer_index: i32,
         image_data: &[u8],
-        offset_alignment: usize,
     ) -> hal::command::Submit<B, hal::Graphics, hal::command::OneShot, hal::command::Primary>
     {
-        //let (image_width, image_height, _, _) = self.kind.dimensions();
         let pos = rect.origin;
         let size = rect.size;
-        let offset = self.upload_buffer.update(device, image_data, offset_alignment);
+        staging_buffer_pool.add(device, image_data);
+        let buffer = staging_buffer_pool.buffer();
         let mut cmd_buffer = cmd_pool.acquire_command_buffer(false);
 
         let range = hal::image::SubresourceRange {
             aspects: hal::format::Aspects::COLOR,
             levels: 0 .. 1,
             layers: layer_index as _ .. (layer_index + 1) as _,
         };
-        if let Some(barrier) = self.core.transit(
-            hal::image::Access::TRANSFER_WRITE,
-            hal::image::Layout::TransferDstOptimal,
-            range.clone(),
-        ) {
+        let mut barriers = Vec::new();
```

Comment:
We should avoid the Vec allocation here and in all other places. Either use `SmallVec` or just make it an iterator that you pass to `transit` regardless of whether it's empty or not.
```rust
let bar_read = buffer.transit(hal::buffer::Access::TRANSFER_READ);
let bar_write = self.core.transit(
                hal::image::Access::TRANSFER_WRITE,
                hal::image::Layout::TransferDstOptimal,
                range.clone(),
            );
let barriers = bar_read.into_iter().chain(bar_write);
```